### ISSUE TITLE
PipeWire support

### ISSUE THREAD

**Issue by etircopyh** [mpv-player/mpv#8569]
[PipeWire](https://gitlab.freedesktop.org/pipewire/pipewire)

Due to pulse backend being quite unstable causing constant desyncs and stuttering I propose to add PipeWire support which is already proven to be more reliable and most importantly have more user-friendly developers.

PipeWire now is under active development. Fedora 34 intent to make it [default.](https://fedoraproject.org/wiki/Changes/DefaultPipeWire) While I understand that PW's API docs is far from completeness this should be at least considered for future implementation. Basically PipeWire can already be used with mpv and any other PulseAudio/JACK app, as all audio from them is routing to PW if one installed. But that still leaves bugs of libpulse and pulse-server in-between. So it would be definitely step forward for Linux audio processing.

[PipeWire Overview](https://gitlab.freedesktop.org/pipewire/pipewire/-/blob/master/doc/overview.md)
[PipeWire Architecture](https://gitlab.freedesktop.org/pipewire/pipewire/-/blob/master/doc/architecture.md)

**Comment by sfan5** [comment]
There has been a proposed PR for a PipeWire audio output last year: #7902
It was not considered for inclusion at the time and IIRC it wasn't quite feature complete either.

> pulse backend being quite unstable causing constant desyncs and stuttering

That's news to me.

**Comment by philipl** [comment]
I thought the official stance from the PipeWire devs was that clients targeting pulse and not writing native integration was a reasonable long term strategy.

**Comment by etircopyh** [comment]
> I thought the official stance from the PipeWire devs was that clients targeting pulse and not writing native integration was a reasonable long term strategy.

Huh? I didn't know it, any reason to think that?

**Comment by philipl** [comment]
https://gitlab.freedesktop.org/pipewire/pipewire/-/wikis/FAQ#what-audio-api-do-you-recommend-to-use

**Comment by etircopyh** [comment]
> [gitlab.freedesktop.org/pipewire/pipewire/-/wikis/FAQ#what-audio-api-do-you-recommend-to-use](https://gitlab.freedesktop.org/pipewire/pipewire/-/wikis/FAQ#what-audio-api-do-you-recommend-to-use)

"for now" doesn't run with your "long term". That's why I said:
> this should be at least considered for future implementation.

**Comment by Oschowa** [comment]
FWIW, a while ago i took the original pull request for the pipewire backend and added some improvement and fixes - mostly format and channel layout negotiation and some error handling. It should be fully functional expect for device enumeration and volume control - which i see no real use for. Anyways, i haven't considered getting it upstreamed, because mpv's ALSA backend works just fine on pipewire, but you can try it if you like: https://github.com/Oschowa/mpv/commit/fddb143282fa74425a8a6f29c9566e51777759d0

**Comment by etircopyh** [comment]
> because mpv's ALSA backend works just fine

Yeah, I don't get this "ALSA works just fine" song

**Comment by mxmilkiib** [comment]
FWIW, [PulseEffects](https://github.com/wwmm/pulseeffects) recently made the shift and has rebranded as "Audio effects for PipeWire applications."

PW has a considerable CPU usage advantage over PA. A friend says they can now leave multiple effects enabled in PE and leave it running in the background now. This helps make apps more stable when they are doing a lot of work.

PipeWire is like JACK but with video. wtay has mentioned full-on video synthesis will be possible (like modular synths are in JACK(+LV2)). It would be very neat to be able to use `mpv` in such a video setup, as a source, or an intermediate node (video in, gets effected and displayed, video out, all with a patchbay/node graph for connections like in [Carla](https://kx.studio/Applications:Carla)).

**Comment by ilya-fedin** [comment]
Yeah, PW is definitely PA done right. I'm really glad that PW finally could be used instead of PA, I've been waiting for this since PW was announced.

**Comment by CounterPillow** [comment]
> PipeWire is like JACK but with video. wtay has mentioned full-on video synthesis will be possible (like modular synths are in JACK(+LV2)). It would be very neat to be able to use mpv in such a video setup

This is more a question of input rather than output. Input would likely be done through an ffmpeg device (like pulse input, JACK input, alsa input and v4l2 input are done right now), not within mpv itself.

**Comment by tinywrkb** [comment]
> all with a patchbay/node graph for connections like in Catia or Carla

See [Helvum](https://gitlab.freedesktop.org/ryuukyu/helvum).

**Comment by tinywrkb** [comment]
@Oschowa your patch works here on my system without issues as far as I can tell.  
It would be nice to have PipeWire output upstreamed as ALSA might not be the best choice when running in a Flatpak sandbox. It's actually broken for me ATM in Flatpak and I'm not sure why, aplay on the host is fine though.  
FTR I running MPV in Flatpak as my OS is immutable, so being able to decouple apps from the base system is useful, and Flatpak let me do this.

**Comment by Oschowa** [comment]
Using ALSA-lib apps in flatpak is currently broken on pipewire, because flatpak forces the use of hte pulseaudio-alsa plug-in (see: https://gitlab.freedesktop.org/pipewire/pipewire/-/issues/860#note_888784)

I've actually found a bug that happens in some rare cases with the pipewire backend patch, so i'll see if i can figure this out first soonish.

**Comment by Markus-N** [comment]
> > pulse backend being quite unstable causing constant desyncs and stuttering
> 
> That's news to me.

I can confirm the desyncs and stuttering:  
With pulse, when I watch videos in mpv or on YouTube and in parallel have a running VM in VirtualBox, I have desyncs and stuttering. The problems stop when I shut down the VM. The OS type and the audio settings of the VM do not make a difference.

During today's update, I removed pulseaudio and installed pipewire instead (plus the pulse compatibility layer provided by the pipewire team), and now the problems are gone.

My OS is Arch Linux.

**Comment by laichiaheng** [comment]
> FWIW, a while ago i took the original pull request for the pipewire backend and added some improvement and fixes - mostly format and channel layout negotiation and some error handling. It should be fully functional expect for device enumeration and volume control - which i see no real use for. Anyways, i haven't considered getting it upstreamed, because mpv's ALSA backend works just fine on pipewire, but you can try it if you like: [Oschowa@fddb143](https://github.com/Oschowa/mpv/commit/fddb143282fa74425a8a6f29c9566e51777759d0)

No, the ALSA backend doesn't work fine, it still has mistimed, delayed and jitter, not like the native PipeWire which has no jitter, mistimed and delayed.

**Comment by eternal-sorrow** [comment]
Mpv doesn't have a native pipewire backend.

**Comment by wm4** [comment]
https://github.com/mpv-player/mpv/commit/fe791 should fix your issue

**Comment by laichiaheng** [comment]
> [fe791](https://github.com/mpv-player/mpv/commit/fe791) should fix your issue

Does it relate to Gnome? I saw it on Plasma with pipewire-pulse too.

**Comment by ojab** [comment]
Feel free to ignore, it's a troll account and this commit not even in mpv.

**Comment by PedroHLC** [comment]
> Mpv doesn't have a native pipewire backend.

Just taking advantage of the bump here to comment, that you can at least use the SDL output (`ao=sdl`) and `SDL_AUDIODRIVER=pipewire` to have something nearer to it... In my case, I needed to build [mpv-full from AUR](aur.archlinux.org/packages/mpv-full) as my distro didn't have the SDL output...

**Comment by etircopyh** [comment]
> > Mpv doesn't have a native pipewire backend.
> 
> Just taking advantage of the bump here to comment, that you can at least use the SDL output (`ao=sdl`) and `SDL_AUDIODRIVER=pipewire` to have something nearer to it... In my case, I needed to build [mpv-full from AUR](aur.archlinux.org/packages/mpv-full) as my distro didn't have the SDL output...

Same with OpenAL (`ao=openal`) with `ALSOFT_DRIVERS=pipewire` variable. Although you'd need to build OpenAL from master and explicitly enable it on mpv's build time.

**Comment by laichiaheng** [comment]
> > > Mpv doesn't have a native pipewire backend.
> > 
> > 
> > Just taking advantage of the bump here to comment, that you can at least use the SDL output (`ao=sdl`) and `SDL_AUDIODRIVER=pipewire` to have something nearer to it... In my case, I needed to build [mpv-full from AUR](aur.archlinux.org/packages/mpv-full) as my distro didn't have the SDL output...
> 
> Same with OpenAL (`ao=openal`) with `ALSOFT_DRIVERS=pipewire` variable. Although you'd need to build OpenAL from master and explicitly enable it on mpv's build time.

It should be written in the manual.

**Comment by etircopyh** [comment]
> It should be written in the manual.

About building OpenAL from master and customizing mpv build? I'm not sure if it's something you put into manual that supposed to describe internal functional. `ao=openal` is already covered, OpenAL specific variables is something to be covered by upstream, which it is.

Information about building third-party software from master is not something you throw into manual, IMO.
